Merge all features related to C# 7 ref features (#795)
dotnet/csharpstandard: standard/statements.md

    ;

local_variable_initializer

>     : 'ref'? expression

Reviewer: To follow the pattern of being explicit in the grammar (see *argument value*) when we can – this is used in comment below

```suggestion
    : expression
    | 'ref' variable_reference
```

    | array_initializer
    ;
```

The *local_variable_type* of a *local_variable_declaration* either directly specifies the type of the variables introduced by the declaration, or indicates with the identifier `var` that the type should be inferred based on an initializer. The type is followed by a list of *local_variable_declarator*s, each of which introduces a new variable. A *local_variable_declarator* consists of an *identifier* that names the variable, optionally followed by an “`=`” token and a *local_variable_initializer* that gives the initial value of the variable. However, it is a compile-time error to omit *local_variable_initializer* from a *local_variable_declarator* for a variable declared `ref` or `ref readonly`.

> The *expression* in a *local_variable_initializer* for a variable declared `ref` or `ref readonly` shall be a variable. It is a compile time error if the scope of the local variable is wider than the *ref_safe_scope* of the *local_variable_initializer* expression (§ref-safe-contexts).

Reviewer: To match the modified grammar above:

```suggestion
A *local_variable_initializer* for a variable declared `ref` or `ref readonly` shall be of the form “`ref` *variable_reference*”. It is a compile time error if the scope of the local variable is wider than the *ref_safe_scope* of the *variable_reference*  (§ref-safe-contexts).
```
Note that the grammar *could* be written so that reference variables can only be initialized by “`ref` *variable_reference*” rather than any of the *local_variable_initializer* forms. Given the expansion of references doing so might reduce repetition (much of it introduced by this PR).

A *local_variable_initializer* for a variable declared `ref` or `ref readonly` shall be of the form “`ref` *variable_reference*”. It is a compile time error if the scope of the local variable is wider than the *ref_safe_scope* of the *variable_reference*  (§ref-safe-contexts).

If *local_variable_declaration* contains `ref readonly`, the *identifier*s being declared are references to variables that are treated as read-only, and their corresponding *local_variable_initializer*s shall each contain `ref`. Otherwise, if *local_variable_declaration* contains `ref` without `readonly`, the *identifier*s being declared are references to variables that shall be writable, and their corresponding *local_variable_initializer* shall each contain `ref`.

> It is a compile-time error to declare a local variable `ref` or `ref readonly` within a method declared with the *method_modifier* `async`.

Reviewer 2: What about iterator methods?

Reviewer 2: What about locals of a ref struct type?

Author: locals of a ref struct type are allowed in iterators, but not async methods. Edited, and added a new paragraph.

In the context of a local variable declaration, the identifier `var` acts as a contextual keyword ([§6.4.4](lexical-structure.md#644-keywords)).When the *local_variable_type* is specified as `var` and no type named `var` is in scope, the declaration is an ***implicitly typed local variable declaration***, whose type is inferred from the type of the associated initializer expression. Implicitly typed local variable declarations are subject to the following restrictions:

- The *local_variable_declaration* cannot include multiple *local_variable_declarator*s.
- The *local_variable_declarator* shall include a *local_variable_initializer*.

> - The *local_variable_initializer* shall be an *expression*, optionally preceded by `ref` or `ref readonly`.

Reviewer 2: Please give an example of a local variable declaration with `ref` or `ref readonly` on the right-hand-side.

Author: Added in the examples later in this clause (starting at line 398)

```ANTLR
foreach_statement
    : 'foreach' '(' ('ref' 'readonly'?)? local_variable_type identifier 'in' 
      expression ')' embedded_statement
    ;
```

> The *local_variable_type* and *Identifier* of a foreach statement declare the ***iteration variable*** of the statement. If the `var` identifier is given as the *local_variable_type*, and no type named `var` is in scope, the iteration variable is said to be an ***implicitly typed iteration variable***, and its type is taken to be the element type of the `foreach` statement, as specified below.

Reviewer 3: >Identifier [](http://example.com/codeflow?start=31&length=10)

Was the change to uppercase "I" intentional?

Reviewer: @cston thanks for spotting this. Lowercase is correct here, ANTLR capitalises lexical rules but this is not one (*Simple_Identifier* is).

```ANTLR
foreach_statement
    : 'foreach' '(' ('ref' 'readonly'?)? local_variable_type identifier 'in' 
      expression ')' embedded_statement
    ;
```

The *local_variable_type* and *identifier* of a foreach statement declare the ***iteration variable*** of the statement. If the `var` identifier is given as the *local_variable_type*, and no type named `var` is in scope, the iteration variable is said to be an ***implicitly typed iteration variable***, and its type is taken to be the element type of the `foreach` statement, as specified below.

> If *foreach_statement* contains both or neither `ref` and `readonly`, the iteration variable denotes a variable that is treated as read-only. Otherwise, if *foreach_statement* contains `ref` without `readonly`, the iteration variable denotes a variable that shall be writable.

Reviewer 4: "If a" or "If the"? I think we need an article of some kind.

> The iteration variable corresponds to a local variable with a scope that extends over the embedded statement. During execution of a `foreach` statement, the iteration variable represents the collection element for which an iteration is currently being performed. If the iteration variable denotes a read-only variable, a compile-time error occurs if the embedded statement attempts to modify it (via assignment or the `++` and `--` operators) or pass it as a `ref` or `out` parameter.

Reviewer 2: The previous paragraph said it is readonly, and this paragraph says that is optional.

I suspect the constraint is that you cannot use the `= ref` operator on the variable.

Author: I think the text is correct, if awkward. I'm making the following suggestion to simplify. The previous paragraph states the conditions under which the variable is a read-only variable. If it's read-only, it can't be modified (by the definition of read-only). So, that sentence is just restating the obvious:

```suggestion
The iteration variable corresponds to a local variable with a scope that extends over the embedded statement. During execution of a `foreach` statement, the iteration variable represents the collection element for which an iteration is currently being performed.
```

Reviewer: In what version of C# is a `ref` foreach iteration variable supported?

Author: Added in 7.3.  Here's an example:

```csharp
    public class C2
    {
        private int[] storage = new int[20];

        public void Initialize()
        {
            int index = 0;
            foreach (ref int i in storage.AsSpan())
            {
                i = index++;
            }
        }
    }
```

Reviewer: Thanks @BillWagner, I'd failed to pin it down.

I think there is a big chunk missing from the `foreach` description – while it defines that the iteration variable maybe a reference variable it doesn’t describe that the type of the *expression* in this case must be one with a ref-returning enumerator. As your example shows arrays do not fit this requirement, you need to create a `Span` which does have such an enumerator.

At a quick read I’m not immediately confident that a ref-foreach will pass the compile-time processing description.

We need to review the description and add at least one example of a ref-foreach.

Reviewer 4: Bill has created #804 to cover this. It definitely needs to be done before C# 7 though.

### 13.10.5 The return statement

The `return` statement returns control to the current caller of the function member in which the return statement appears, optionally returning a value or a *variable_reference* (§10.5).

```ANTLR
return_statement
    : 'return' ';'

>     | 'return' 'ref'? expression ';'

Reviewer: See general review comments

```suggestion
    | 'return' expression ';'
    | 'return' 'ref' variable_reference ';'
```

    ;
```

A *return_statement* without *expression* is called a ***return-no-value***; one containing `ref` *expression* is called a ***return-by-ref***; and one containing only *expression* is called a ***return-by-value***.

It is a compile-time error to use a return-no-value from a method declared as being returns-by-value or returns-by-ref ([§15.6.1](classes.md#1561-general)).

It is a compile-time error to use a return-by-ref from a method declared as being returns-no-value or returns-by-value.

It is a compile-time error to use a return-by-value from a method declared as being returns-no-value or returns-by-ref.

It is a compile-time error to use a return-by-ref if *expression* is not a *variable_reference* or is a reference to a variable whose *ref_safe_scope* is not *calling_method* (§ref-safe-contexts).

It is a compile-time error to use a return-by-ref from a method declared with the *method_modifier* `async`.

> A function member is said to ***compute a value*** if it is a method with a returns-by-value method ([§15.6.11](classes.md#15611-method-body)), a returns-by-value `get` accessor of a property or indexer, or a user-defined operator. Function members that are returns-no-value do not compute a value or are methods with the effective return type `void`, `set` accessors of properties and indexers, `add` and `remove` accessors of event, instance constructors, static constructors and finalizers. Function members that are returns-by-ref do not compute a value.

Reviewer 3: >or are methods [](http://example.com/codeflow?start=298&length=14)

Should this be "_and_ are methods ..."?

> For a return-by-value, an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) shall exist from the type of *expression* to the effective return type ([§15.6.11](classes.md#15611-method-body)) of the containing function member. For a return-by-ref, and identity conversion ([§10.2.2](conversions.md#1022-identity-conversion)) shall exist between the type of *expression* and  the effective return type of the containing function member.

Reviewer 3: >and identity conversion [](http://example.com/codeflow?start=267&length=23)

"_an_ identity conversion ..."?

For a return-by-value, an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) shall exist from the type of *expression* to the effective return type ([§15.6.11](classes.md#15611-method-body)) of the containing function member. For a return-by-ref, an identity conversion ([§10.2.2](conversions.md#1022-identity-conversion)) shall exist between the type of *expression* and  the effective return type of the containing function member.

`return` statements can also be used in the body of anonymous function expressions ([§12.19](expressions.md#1219-anonymous-function-expressions)), and participate in determining which conversions exist for those functions ([§10.7.1](conversions.md#1071-general)).

It is a compile-time error for a `return` statement to appear in a `finally` block ([§13.11](statements.md#1311-the-try-statement)).

A `return` statement is executed as follows:

> - For a return-by-value, *expression* is evaluated and its value is converted to the effective return type of the containing function by an implicit conversion. The result of the conversion becomes the result value produced by the function. For a return-by-ref, a reference to the *variable_reference* designated by *expression* becomes the result produced by the function. That result is a variable. If the enclosing method’s return-by-ref includes `readonly`, the resulting variable is read-only.

Reviewer 2: Are nested expressions inside the expression of a return-by-ref evaluated?

Reviewer: Yes. Well at least as much as normal C# evaluation rules require them to be to produce a variable reference, e.g. conditional expressions only evaluate two of their three arguments by definition. Why would it be any different here, and why would C# evaluation rules need to be mentioned? What am I missing?

Reviewer 4: Something like `return ref array[MethodCall()]`?

Reviewer 4: Bill has created #805 as a clarity bug, to be fixed post C# 7.
